{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualising the NK cluster\n",
    "\n",
    "## finding NK gene candidates\n",
    "\n",
    "We compiled a list of homeobox genes in the NK family from the [Aase-Remedios _et al._\n",
    "(2023)](https://doi.org/10.1093/molbev/msad239) analysis of the spider Homeobox gene repertoire. We\n",
    "used this list to scan against the _P. litorale_ draft genome using `mmseqs`:\n",
    "\n",
    "```bash\n",
    "cd /lisc/scratch/zoology/pycnogonum/genome/draft/nkx\n",
    "module load mmseqs2\n",
    "M8FORMAT=\"query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qlen\"\n",
    "mmseqs easy-search nk_chelicerates.fa ../draft.fasta nk_genomic.m8 tmp --format-output $M8FORMAT --threads 4\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we'll have to read the GFF file so that we can locate the loci that overlap with our genomic\n",
    "hits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from sklearn.cluster import DBSCAN\n",
    "from sklearn.datasets import make_blobs\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.patches import ConnectionPatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gff(loc):\n",
    "    gff = pd.read_csv(loc, sep=\"\\t\", header=None, skiprows=4)\n",
    "    gff_columns = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "    gff.columns = gff_columns\n",
    "    return gff\n",
    "\n",
    "# read file, name columns\n",
    "def read_aln(m8, id_sep=None):\n",
    "    nk = pd.read_csv(m8, sep=\"\\t\", header=None)\n",
    "    m8_columns = ['query', 'target', 'seq_id', 'ali_len', 'no_mism', 'no_go',\n",
    "                'q_start', 'q_end', 't_start', 't_end', 'eval', 'bit', 'qlen']\n",
    "    nk.columns = m8_columns\n",
    "    # trim the query to just the ID\n",
    "    if id_sep is not None:\n",
    "        nk[\"query\"] = nk[\"query\"].str.split(id_sep).str[1]\n",
    "    return nk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_loc = \"/Volumes/project/pycnogonum/paper/zenodo/results/merged_sorted_named_dedup_flagged.gff3\"\n",
    "tmp = read_gff(gff_loc)\n",
    "gff = tmp[tmp['type'] == 'gene'].copy()\n",
    "gff['gene'] = gff['attributes'].str.split(\"ID=\").str[1].str.split(\";\").str[0]\n",
    "del tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pseudochrom_1</td>\n",
       "      <td>GeneMark.hmm3</td>\n",
       "      <td>gene</td>\n",
       "      <td>119</td>\n",
       "      <td>5102</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g1;gene=Uncharacterised protein g1;gene_id=g1</td>\n",
       "      <td>g1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>pseudochrom_1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>20074</td>\n",
       "      <td>33326</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.1;function=Major Facilitator Superfamily...</td>\n",
       "      <td>PB.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>pseudochrom_1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>33362</td>\n",
       "      <td>39771</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.2;gene=Uncharacterised protein PB.2;gene...</td>\n",
       "      <td>PB.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>pseudochrom_1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>50782</td>\n",
       "      <td>71089</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.3;function=ribose phosphate diphosphokin...</td>\n",
       "      <td>PB.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>pseudochrom_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>63017</td>\n",
       "      <td>66568</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g5;function=Bifunctional DNA N-glycosylase ...</td>\n",
       "      <td>g5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761020</th>\n",
       "      <td>scaffold_12318</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>1232</td>\n",
       "      <td>15217</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8896;gene=Uncharacterised protein PB.889...</td>\n",
       "      <td>PB.8896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761038</th>\n",
       "      <td>scaffold_12330</td>\n",
       "      <td>AGAT</td>\n",
       "      <td>gene</td>\n",
       "      <td>6041</td>\n",
       "      <td>13951</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=r2_g4229;gene=Uncharacterised protein r2_g4...</td>\n",
       "      <td>r2_g4229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761046</th>\n",
       "      <td>scaffold_12335</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>1064</td>\n",
       "      <td>1939</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8840;gene=Uncharacterised protein PB.884...</td>\n",
       "      <td>PB.8840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761050</th>\n",
       "      <td>scaffold_12338</td>\n",
       "      <td>tRNAscan-SE</td>\n",
       "      <td>gene</td>\n",
       "      <td>19482</td>\n",
       "      <td>19554</td>\n",
       "      <td>29.5</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=tRNA905;gene_id=scaffold_12338.tRNA1-TyrGTA...</td>\n",
       "      <td>tRNA905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761052</th>\n",
       "      <td>scaffold_12360</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>4755</td>\n",
       "      <td>34656</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8875;function=zinc ion binding. It is in...</td>\n",
       "      <td>PB.8875</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16459 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seqid         source  type  start    end score strand phase  \\\n",
       "0        pseudochrom_1  GeneMark.hmm3  gene    119   5102     .      +     .   \n",
       "14       pseudochrom_1         PacBio  gene  20074  33326     .      -     .   \n",
       "67       pseudochrom_1         PacBio  gene  33362  39771     .      +     .   \n",
       "117      pseudochrom_1         PacBio  gene  50782  71089     .      -     .   \n",
       "259      pseudochrom_1       AUGUSTUS  gene  63017  66568     .      +     .   \n",
       "...                ...            ...   ...    ...    ...   ...    ...   ...   \n",
       "761020  scaffold_12318         PacBio  gene   1232  15217     .      -     .   \n",
       "761038  scaffold_12330           AGAT  gene   6041  13951     .      +     .   \n",
       "761046  scaffold_12335         PacBio  gene   1064   1939     .      +     .   \n",
       "761050  scaffold_12338    tRNAscan-SE  gene  19482  19554  29.5      -     .   \n",
       "761052  scaffold_12360         PacBio  gene   4755  34656     .      +     .   \n",
       "\n",
       "                                               attributes      gene  \n",
       "0        ID=g1;gene=Uncharacterised protein g1;gene_id=g1        g1  \n",
       "14      ID=PB.1;function=Major Facilitator Superfamily...      PB.1  \n",
       "67      ID=PB.2;gene=Uncharacterised protein PB.2;gene...      PB.2  \n",
       "117     ID=PB.3;function=ribose phosphate diphosphokin...      PB.3  \n",
       "259     ID=g5;function=Bifunctional DNA N-glycosylase ...        g5  \n",
       "...                                                   ...       ...  \n",
       "761020  ID=PB.8896;gene=Uncharacterised protein PB.889...   PB.8896  \n",
       "761038  ID=r2_g4229;gene=Uncharacterised protein r2_g4...  r2_g4229  \n",
       "761046  ID=PB.8840;gene=Uncharacterised protein PB.884...   PB.8840  \n",
       "761050  ID=tRNA905;gene_id=scaffold_12338.tRNA1-TyrGTA...   tRNA905  \n",
       "761052  ID=PB.8875;function=zinc ion binding. It is in...   PB.8875  \n",
       "\n",
       "[16459 rows x 10 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "nk = read_aln('/Volumes/scratch/pycnogonum/genome/draft/nkx/nk_genomic.m8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "nk['species'] = nk['query'].str.split(\"_\").str[:2].str.join(\" \")\n",
    "nk['symbol'] = nk['query'].str.split(\"_\").str[2:].str.join(\"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hit_to_locus(hit, gff):\n",
    "    same_chrom = gff['seqid'] == hit['target']\n",
    "    within_borders = (gff['start'] <= hit['t_start']) & (gff['end'] >= hit['t_end'])\n",
    "\n",
    "    overlap = gff[same_chrom & within_borders]\n",
    "    return overlap['gene'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "hits = nk[nk['eval'] < 1e-20].copy()\n",
    "hits['gene'] = hits.apply(hit_to_locus, axis=1, gff=gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "g8689      84\n",
       "PB.3341    83\n",
       "PB.3347    83\n",
       "PB.1018    64\n",
       "g10110     64\n",
       "PB.1017    64\n",
       "PB.1005    41\n",
       "g6828      39\n",
       "PB.1003    36\n",
       "g8691      31\n",
       "g6762      31\n",
       "PB.8153    21\n",
       "PB.7934    20\n",
       "g1625       6\n",
       "g1627       6\n",
       "g10112      4\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits['gene'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "candidates = hits['gene'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "plit_peptides = {}\n",
    "with open('/Volumes/project/pycnogonum/genome/draft/transcripts.fa.transdecoder.pep', 'r') as f:\n",
    "    lines = f.readlines()\n",
    "    for i, line in enumerate(lines):\n",
    "        if line.startswith(\">\"):\n",
    "            gene = line.split(\" \")[0][1:]\n",
    "            plit_peptides[gene] = \"\"\n",
    "        else:\n",
    "            plit_peptides[gene] += line.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "candidate_sequences = {}\n",
    "for gene in plit_peptides:\n",
    "    for c in candidates:\n",
    "        if gene.startswith(c):\n",
    "            candidate_sequences[gene] = plit_peptides[gene]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Volumes/scratch/pycnogonum/genome/draft/nkx/nk_candidates.fa', 'w') as f:\n",
    "    for gene, seq in candidate_sequences.items():\n",
    "        f.write(f\">{gene}\\n{seq}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can concatenate all the sequences and align them.\n",
    "\n",
    "```bash\n",
    "cat nk_chelicerates.fa > nk.fa\n",
    "cat nk_candidates.fa >> nk.fa\n",
    "```\n",
    "\n",
    "First, quick-and-dirty alignment so we can just cut out the homeodomain regions from the _P.\n",
    "litorale_ sequences:\n",
    "\n",
    "```bash\n",
    "mafft --thread 4 nk.fa > nk_aligned.fasta\n",
    "```\n",
    "\n",
    "Now we can extract the homeodomain regions. This was done manually in Jalview, by deleting columns\n",
    "left and right of the homeobox region. Two _P. litorale_ isoforms that broke up the homeobox domain\n",
    "were excluded completely. Additionally, the _Hhex_ sequences were misaligned, which led to a break\n",
    "in the homeobox domain block. We trimmed up to the beginning of the homeobox domain and included the\n",
    "break, hoping that a realignment would fix this.\n",
    "\n",
    "The resulting trimmed alignment was saved as `homeobox.fasta` and\n",
    "re-aligned with MAFFT in L-INS-i mode (Probably most accurate, a little bit slower):\n",
    "\n",
    "```bash\n",
    "mafft --thread 4 --maxiterate 1000 --localpair homeobox.fasta > homeobox_aligned.fasta\n",
    "```\n",
    "\n",
    "The re-alignment improved the alignment quality, and we could safely trim the _P. litorale_\n",
    "positions that mapped to the misplaced _Hhex_ parts. The resulting alignment was input to IQTREE2.\n",
    "\n",
    "```bash\n",
    "> module load iqtree\n",
    "> /usr/bin/time iqtree2 -s ./homeobox_aligned.fasta -B 1000 -T 4\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ascc24",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
