{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualising the NK cluster\n",
    "\n",
    "## finding NK gene candidates\n",
    "\n",
    "We compiled a list of homeobox genes in the NK family from the [Aase-Remedios _et al._\n",
    "(2023)](https://doi.org/10.1093/molbev/msad239) analysis of the spider Homeobox gene repertoire. We\n",
    "used this list to scan against the _P. litorale_ draft genome using `mmseqs`:\n",
    "\n",
    "```bash\n",
    "cd /lisc/scratch/zoology/pycnogonum/genome/draft/nkx\n",
    "module load mmseqs2\n",
    "M8FORMAT=\"query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qlen\"\n",
    "mmseqs easy-search nk.fa ../draft.fasta nk_genomic.m8 tmp --format-output $M8FORMAT --threads 4\n",
    "```\n",
    "\n",
    "Similarly, we used the same list to scan against the deeply sequenced _P. litorale_ transcriptomes:\n",
    "\n",
    "```bash\n",
    "for pep in ../transcriptome/*/Trinity.fasta.transdecoder.pep; do\n",
    "    stage=$(echo $pep | cut -d\"/\" -f3);\n",
    "    mmseqs easy-search nk.fa $pep nk_$stage.m8 tmp --format-output $M8FORMAT --threads 4\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we'll have to read the GFF file so that we can locate the loci that overlap with our genomic\n",
    "hits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from sklearn.cluster import DBSCAN\n",
    "from sklearn.datasets import make_blobs\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.patches import ConnectionPatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gff(loc):\n",
    "    gff = pd.read_csv(loc, sep=\"\\t\", header=None, skiprows=4)\n",
    "    gff_columns = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "    gff.columns = gff_columns\n",
    "    return gff\n",
    "\n",
    "# read file, name columns\n",
    "def read_aln(m8, id_sep=None):\n",
    "    hox = pd.read_csv(m8, sep=\"\\t\", header=None)\n",
    "    m8_columns = ['query', 'target', 'seq_id', 'ali_len', 'no_mism', 'no_go',\n",
    "                'q_start', 'q_end', 't_start', 't_end', 'eval', 'bit', 'qlen']\n",
    "    hox.columns = m8_columns\n",
    "    # trim the query to just the ID\n",
    "    if id_sep is not None:\n",
    "        hox[\"query\"] = hox[\"query\"].str.split(id_sep).str[1]\n",
    "    return hox"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_loc = \"/Volumes/project/pycnogonum/paper/zenodo/results/merged_sorted_named_dedup_flagged.gff3\"\n",
    "tmp = read_gff(gff_loc)\n",
    "gff = tmp[tmp['type'] == 'gene'].copy()\n",
    "gff['gene'] = gff['attributes'].str.split(\"ID=\").str[1].str.split(\";\").str[0]\n",
    "del tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pseudochrom_1</td>\n",
       "      <td>GeneMark.hmm3</td>\n",
       "      <td>gene</td>\n",
       "      <td>119</td>\n",
       "      <td>5102</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g1;gene=Uncharacterised protein g1;gene_id=g1</td>\n",
       "      <td>g1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>pseudochrom_1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>20074</td>\n",
       "      <td>33326</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.1;function=Major Facilitator Superfamily...</td>\n",
       "      <td>PB.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>pseudochrom_1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>33362</td>\n",
       "      <td>39771</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.2;gene=Uncharacterised protein PB.2;gene...</td>\n",
       "      <td>PB.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>pseudochrom_1</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>50782</td>\n",
       "      <td>71089</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.3;function=ribose phosphate diphosphokin...</td>\n",
       "      <td>PB.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>pseudochrom_1</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>63017</td>\n",
       "      <td>66568</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g5;function=Bifunctional DNA N-glycosylase ...</td>\n",
       "      <td>g5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761020</th>\n",
       "      <td>scaffold_12318</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>1232</td>\n",
       "      <td>15217</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8896;gene=Uncharacterised protein PB.889...</td>\n",
       "      <td>PB.8896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761038</th>\n",
       "      <td>scaffold_12330</td>\n",
       "      <td>AGAT</td>\n",
       "      <td>gene</td>\n",
       "      <td>6041</td>\n",
       "      <td>13951</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=r2_g4229;gene=Uncharacterised protein r2_g4...</td>\n",
       "      <td>r2_g4229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761046</th>\n",
       "      <td>scaffold_12335</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>1064</td>\n",
       "      <td>1939</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8840;gene=Uncharacterised protein PB.884...</td>\n",
       "      <td>PB.8840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761050</th>\n",
       "      <td>scaffold_12338</td>\n",
       "      <td>tRNAscan-SE</td>\n",
       "      <td>gene</td>\n",
       "      <td>19482</td>\n",
       "      <td>19554</td>\n",
       "      <td>29.5</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=tRNA905;gene_id=scaffold_12338.tRNA1-TyrGTA...</td>\n",
       "      <td>tRNA905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>761052</th>\n",
       "      <td>scaffold_12360</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>4755</td>\n",
       "      <td>34656</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8875;function=zinc ion binding. It is in...</td>\n",
       "      <td>PB.8875</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16459 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seqid         source  type  start    end score strand phase  \\\n",
       "0        pseudochrom_1  GeneMark.hmm3  gene    119   5102     .      +     .   \n",
       "14       pseudochrom_1         PacBio  gene  20074  33326     .      -     .   \n",
       "67       pseudochrom_1         PacBio  gene  33362  39771     .      +     .   \n",
       "117      pseudochrom_1         PacBio  gene  50782  71089     .      -     .   \n",
       "259      pseudochrom_1       AUGUSTUS  gene  63017  66568     .      +     .   \n",
       "...                ...            ...   ...    ...    ...   ...    ...   ...   \n",
       "761020  scaffold_12318         PacBio  gene   1232  15217     .      -     .   \n",
       "761038  scaffold_12330           AGAT  gene   6041  13951     .      +     .   \n",
       "761046  scaffold_12335         PacBio  gene   1064   1939     .      +     .   \n",
       "761050  scaffold_12338    tRNAscan-SE  gene  19482  19554  29.5      -     .   \n",
       "761052  scaffold_12360         PacBio  gene   4755  34656     .      +     .   \n",
       "\n",
       "                                               attributes      gene  \n",
       "0        ID=g1;gene=Uncharacterised protein g1;gene_id=g1        g1  \n",
       "14      ID=PB.1;function=Major Facilitator Superfamily...      PB.1  \n",
       "67      ID=PB.2;gene=Uncharacterised protein PB.2;gene...      PB.2  \n",
       "117     ID=PB.3;function=ribose phosphate diphosphokin...      PB.3  \n",
       "259     ID=g5;function=Bifunctional DNA N-glycosylase ...        g5  \n",
       "...                                                   ...       ...  \n",
       "761020  ID=PB.8896;gene=Uncharacterised protein PB.889...   PB.8896  \n",
       "761038  ID=r2_g4229;gene=Uncharacterised protein r2_g4...  r2_g4229  \n",
       "761046  ID=PB.8840;gene=Uncharacterised protein PB.884...   PB.8840  \n",
       "761050  ID=tRNA905;gene_id=scaffold_12338.tRNA1-TyrGTA...   tRNA905  \n",
       "761052  ID=PB.8875;function=zinc ion binding. It is in...   PB.8875  \n",
       "\n",
       "[16459 rows x 10 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "nk = read_aln('/Volumes/scratch/pycnogonum/genome/draft/nkx/nk_genomic.m8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hit_to_locus(hit, gff):\n",
    "    same_chrom = gff['seqid'] == hit['target']\n",
    "    within_borders = (gff['start'] <= hit['t_start']) & (gff['end'] >= hit['t_end'])\n",
    "\n",
    "    overlap = gff[same_chrom & within_borders]\n",
    "    return overlap['gene'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "hits = nk[nk['eval'] < 1e-20].copy()\n",
    "hits['gene'] = hits.apply(hit_to_locus, axis=1, gff=gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "g8689      84\n",
       "PB.3341    83\n",
       "PB.3347    83\n",
       "PB.1005    41\n",
       "PB.1003    36\n",
       "g6762      31\n",
       "PB.7934    20\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "best_nk = {}\n",
    "\n",
    "deep_dev_transcriptomes = [\"INSTAR1\", \"INSTAR3\", \"INSTAR5\", \"JUV1\", \"EMBRYO3\", \"INSTAR2\", \"INSTAR4\", \"INSTAR6\", \"SUBADULT\"]\n",
    "\n",
    "for stage in deep_dev_transcriptomes:\n",
    "    nk_txome_loc = f'/Volumes/scratch/pycnogonum/genome/draft/nkx/nk_{stage}.m8'\n",
    "    nk = read_aln(nk_txome_loc)\n",
    "\n",
    "    top_hits = nk[nk['eval'] < 1e-20]\n",
    "    best_nk[stage] = top_hits['target'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'INSTAR1': array(['TRINITY_DN20982_c2_g1_i1.p1', 'TRINITY_DN1632_c8_g1_i1.p1',\n",
       "        'TRINITY_DN3973_c1_g1_i3.p1', 'TRINITY_DN3973_c1_g1_i4.p1',\n",
       "        'TRINITY_DN3973_c1_g1_i2.p1', 'TRINITY_DN3973_c1_g1_i5.p1',\n",
       "        'TRINITY_DN98526_c0_g2_i1.p1', 'TRINITY_DN20982_c0_g1_i1.p1',\n",
       "        'TRINITY_DN3014_c2_g1_i1.p1', 'TRINITY_DN191_c0_g3_i1.p1'],\n",
       "       dtype=object),\n",
       " 'INSTAR3': array(['TRINITY_DN77411_c0_g1_i1.p1', 'TRINITY_DN1357_c1_g1_i1.p1',\n",
       "        'TRINITY_DN2510_c0_g1_i2.p1', 'TRINITY_DN2510_c0_g1_i4.p1',\n",
       "        'TRINITY_DN2510_c0_g1_i1.p1', 'TRINITY_DN2510_c0_g1_i3.p1',\n",
       "        'TRINITY_DN71961_c0_g1_i1.p1', 'TRINITY_DN4683_c3_g1_i4.p1',\n",
       "        'TRINITY_DN4683_c3_g1_i7.p1', 'TRINITY_DN954_c0_g1_i1.p1',\n",
       "        'TRINITY_DN2546_c0_g1_i10.p1', 'TRINITY_DN2546_c0_g1_i3.p1',\n",
       "        'TRINITY_DN2546_c0_g1_i6.p1', 'TRINITY_DN2546_c0_g1_i1.p1',\n",
       "        'TRINITY_DN2546_c0_g1_i11.p1', 'TRINITY_DN2546_c0_g1_i12.p1',\n",
       "        'TRINITY_DN2546_c0_g1_i7.p1', 'TRINITY_DN3779_c0_g2_i1.p1'],\n",
       "       dtype=object),\n",
       " 'INSTAR5': array(['TRINITY_DN4620_c0_g1_i1.p1', 'TRINITY_DN4319_c0_g1_i1.p1',\n",
       "        'TRINITY_DN57221_c0_g1_i1.p1', 'TRINITY_DN57221_c0_g1_i3.p1',\n",
       "        'TRINITY_DN57221_c0_g1_i5.p1', 'TRINITY_DN57221_c0_g1_i2.p1',\n",
       "        'TRINITY_DN57221_c0_g1_i4.p1', 'TRINITY_DN102_c2_g1_i2.p1',\n",
       "        'TRINITY_DN9621_c0_g1_i1.p1', 'TRINITY_DN9621_c0_g1_i10.p1',\n",
       "        'TRINITY_DN9621_c0_g1_i11.p1', 'TRINITY_DN9621_c0_g1_i12.p1',\n",
       "        'TRINITY_DN9621_c0_g1_i13.p1', 'TRINITY_DN9621_c0_g1_i14.p1',\n",
       "        'TRINITY_DN9621_c0_g1_i15.p1', 'TRINITY_DN9621_c0_g1_i4.p1',\n",
       "        'TRINITY_DN9621_c0_g1_i6.p1', 'TRINITY_DN9621_c0_g1_i7.p1',\n",
       "        'TRINITY_DN9621_c0_g1_i9.p1', 'TRINITY_DN4790_c0_g1_i1.p1',\n",
       "        'TRINITY_DN397_c3_g1_i1.p1', 'TRINITY_DN397_c3_g1_i2.p1',\n",
       "        'TRINITY_DN4698_c0_g1_i1.p1', 'TRINITY_DN102_c2_g1_i1.p1'],\n",
       "       dtype=object),\n",
       " 'JUV1': array(['TRINITY_DN16894_c0_g1_i1.p1', 'TRINITY_DN7635_c0_g1_i1.p1',\n",
       "        'TRINITY_DN17973_c0_g1_i1.p1', 'TRINITY_DN17973_c0_g1_i4.p1',\n",
       "        'TRINITY_DN17973_c0_g1_i2.p1', 'TRINITY_DN17973_c0_g1_i3.p1',\n",
       "        'TRINITY_DN2567_c1_g1_i1.p1', 'TRINITY_DN2567_c1_g1_i3.p1',\n",
       "        'TRINITY_DN2567_c1_g1_i4.p1', 'TRINITY_DN2567_c1_g1_i5.p1',\n",
       "        'TRINITY_DN2567_c1_g1_i6.p1', 'TRINITY_DN8195_c0_g1_i1.p1',\n",
       "        'TRINITY_DN8195_c0_g1_i2.p1', 'TRINITY_DN8195_c0_g1_i4.p1',\n",
       "        'TRINITY_DN9482_c0_g1_i1.p1', 'TRINITY_DN2719_c0_g1_i1.p1',\n",
       "        'TRINITY_DN2719_c0_g1_i14.p1', 'TRINITY_DN2719_c0_g1_i2.p1',\n",
       "        'TRINITY_DN2719_c0_g1_i9.p1', 'TRINITY_DN9897_c0_g1_i1.p1',\n",
       "        'TRINITY_DN9897_c0_g1_i2.p1'], dtype=object),\n",
       " 'EMBRYO3': array(['TRINITY_DN5303_c2_g1_i1.p1', 'TRINITY_DN20522_c0_g1_i1.p1',\n",
       "        'TRINITY_DN20522_c0_g1_i10.p1', 'TRINITY_DN20522_c0_g1_i2.p1',\n",
       "        'TRINITY_DN20522_c0_g1_i3.p1', 'TRINITY_DN20522_c0_g1_i5.p1',\n",
       "        'TRINITY_DN20522_c0_g1_i6.p1', 'TRINITY_DN20522_c0_g1_i7.p1',\n",
       "        'TRINITY_DN20522_c0_g1_i8.p1', 'TRINITY_DN4289_c0_g1_i3.p1',\n",
       "        'TRINITY_DN4289_c0_g1_i1.p1', 'TRINITY_DN4289_c0_g1_i2.p1',\n",
       "        'TRINITY_DN1966_c0_g1_i1.p1', 'TRINITY_DN1966_c0_g1_i2.p1',\n",
       "        'TRINITY_DN1966_c0_g1_i3.p1', 'TRINITY_DN5303_c0_g1_i1.p1',\n",
       "        'TRINITY_DN45791_c0_g2_i1.p1', 'TRINITY_DN6478_c3_g1_i1.p1',\n",
       "        'TRINITY_DN6478_c3_g1_i2.p1'], dtype=object),\n",
       " 'INSTAR2': array(['TRINITY_DN11408_c0_g5_i1.p1', 'TRINITY_DN4562_c1_g1_i1.p1',\n",
       "        'TRINITY_DN1403_c0_g1_i1.p1', 'TRINITY_DN1403_c0_g1_i3.p1',\n",
       "        'TRINITY_DN1403_c0_g1_i2.p1', 'TRINITY_DN6715_c1_g1_i1.p1',\n",
       "        'TRINITY_DN3894_c1_g1_i1.p1', 'TRINITY_DN3894_c1_g1_i2.p1',\n",
       "        'TRINITY_DN3894_c1_g1_i3.p1', 'TRINITY_DN3894_c1_g1_i4.p1',\n",
       "        'TRINITY_DN3894_c1_g1_i6.p1', 'TRINITY_DN3894_c1_g1_i7.p1',\n",
       "        'TRINITY_DN9957_c0_g1_i1.p1', 'TRINITY_DN285_c1_g1_i2.p1'],\n",
       "       dtype=object),\n",
       " 'INSTAR4': array(['TRINITY_DN3615_c1_g1_i1.p1', 'TRINITY_DN156_c2_g1_i1.p1',\n",
       "        'TRINITY_DN1575_c1_g1_i1.p1', 'TRINITY_DN1575_c1_g1_i2.p1',\n",
       "        'TRINITY_DN8609_c0_g1_i1.p1', 'TRINITY_DN1792_c2_g2_i1.p1',\n",
       "        'TRINITY_DN505_c3_g4_i1.p1', 'TRINITY_DN22862_c0_g1_i4.p1',\n",
       "        'TRINITY_DN22862_c0_g1_i19.p1', 'TRINITY_DN22862_c0_g1_i5.p1',\n",
       "        'TRINITY_DN22862_c0_g1_i11.p1', 'TRINITY_DN22862_c0_g1_i12.p1',\n",
       "        'TRINITY_DN22862_c0_g1_i14.p1', 'TRINITY_DN22862_c0_g1_i15.p1',\n",
       "        'TRINITY_DN22862_c0_g1_i18.p1', 'TRINITY_DN22862_c0_g1_i20.p1',\n",
       "        'TRINITY_DN22862_c0_g1_i3.p1', 'TRINITY_DN61964_c0_g1_i1.p1'],\n",
       "       dtype=object),\n",
       " 'INSTAR6': array(['TRINITY_DN14039_c1_g1_i1.p1', 'TRINITY_DN5474_c0_g1_i1.p1',\n",
       "        'TRINITY_DN5474_c0_g1_i2.p1', 'TRINITY_DN643_c1_g1_i1.p1',\n",
       "        'TRINITY_DN643_c1_g1_i5.p1', 'TRINITY_DN643_c1_g1_i2.p1',\n",
       "        'TRINITY_DN643_c1_g1_i3.p1', 'TRINITY_DN63142_c0_g1_i1.p1',\n",
       "        'TRINITY_DN63142_c0_g1_i2.p1', 'TRINITY_DN63142_c0_g1_i5.p1',\n",
       "        'TRINITY_DN63142_c0_g1_i7.p1', 'TRINITY_DN3687_c0_g1_i5.p1',\n",
       "        'TRINITY_DN3687_c0_g1_i1.p1', 'TRINITY_DN1040_c1_g1_i1.p1',\n",
       "        'TRINITY_DN293_c6_g1_i1.p1', 'TRINITY_DN293_c6_g1_i2.p1',\n",
       "        'TRINITY_DN10836_c0_g1_i1.p1'], dtype=object),\n",
       " 'SUBADULT': array(['TRINITY_DN6208_c1_g1_i1.p1', 'TRINITY_DN7960_c1_g1_i1.p1',\n",
       "        'TRINITY_DN7960_c1_g1_i2.p1', 'TRINITY_DN653_c0_g1_i5.p1',\n",
       "        'TRINITY_DN653_c0_g1_i1.p1', 'TRINITY_DN653_c0_g1_i4.p1',\n",
       "        'TRINITY_DN653_c0_g1_i6.p1', 'TRINITY_DN653_c0_g1_i2.p1',\n",
       "        'TRINITY_DN653_c0_g1_i3.p1', 'TRINITY_DN4696_c0_g1_i1.p1',\n",
       "        'TRINITY_DN4696_c0_g1_i2.p1', 'TRINITY_DN1610_c10_g1_i1.p1',\n",
       "        'TRINITY_DN3581_c4_g1_i3.p1', 'TRINITY_DN3581_c4_g1_i1.p1',\n",
       "        'TRINITY_DN3581_c4_g1_i2.p1', 'TRINITY_DN35158_c0_g1_i1.p1',\n",
       "        'TRINITY_DN35158_c0_g1_i2.p1'], dtype=object)}"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_nk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ascc24",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
