{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualising the SINE cluster\n",
    "\n",
    "## finding SINE gene candidates\n",
    "\n",
    "We compiled a list of homeobox genes in the SINE family from the [Aase-Remedios _et al._\n",
    "(2023)](https://doi.org/10.1093/molbev/msad239) analysis of the spider Homeobox gene repertoire. We\n",
    "used this list to scan against the _P. litorale_ draft genome using `mmseqs`:\n",
    "\n",
    "```bash\n",
    "cd /lisc/scratch/zoology/pycnogonum/genome/draft/sine\n",
    "module load mmseqs2\n",
    "M8FORMAT=\"query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits,qlen\"\n",
    "mmseqs easy-search sine_chelicerates.fa ../draft.fasta sine_genomic.m8 tmp --format-output $M8FORMAT --threads 4\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we'll have to read the GFF file so that we can locate the loci that overlap with our genomic\n",
    "hits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from sklearn.cluster import DBSCAN\n",
    "from sklearn.datasets import make_blobs\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.patches import ConnectionPatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gff(loc):\n",
    "    gff = pd.read_csv(loc, sep=\"\\t\", header=None, skiprows=4)\n",
    "    gff_columns = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "    gff.columns = gff_columns\n",
    "    return gff\n",
    "\n",
    "# read file, name columns\n",
    "def read_aln(m8, id_sep=None):\n",
    "    sine = pd.read_csv(m8, sep=\"\\t\", header=None)\n",
    "    m8_columns = ['query', 'target', 'seq_id', 'ali_len', 'no_mism', 'no_go',\n",
    "                'q_start', 'q_end', 't_start', 't_end', 'eval', 'bit', 'qlen']\n",
    "    sine.columns = m8_columns\n",
    "    # trim the query to just the ID\n",
    "    if id_sep is not None:\n",
    "        sine[\"query\"] = sine[\"query\"].str.split(id_sep).str[1]\n",
    "    return sine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_loc = \"/Volumes/project/pycnogonum/paper/zenodo/results/merged_sorted_named_dedup_flagged.gff3\"\n",
    "tmp = read_gff(gff_loc)\n",
    "gff = tmp[tmp['type'] == 'gene'].copy()\n",
    "gff['gene'] = gff['attributes'].str.split(\"ID=\").str[1].str.split(\";\").str[0]\n",
    "del tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "sine = read_aln('/Volumes/scratch/pycnogonum/genome/draft/sine/sine_genomic.m8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "sine['species'] = sine['query'].str.split(\"_\").str[:2].str.join(\" \")\n",
    "sine['symbol'] = sine['query'].str.split(\"_\").str[2:].str.join(\"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hit_to_locus(hit, gff):\n",
    "    same_chrom = gff['seqid'] == hit['target']\n",
    "    within_borders = (gff['start'] <= hit['t_start']) & (gff['end'] >= hit['t_end'])\n",
    "\n",
    "    overlap = gff[same_chrom & within_borders]\n",
    "    return overlap['gene'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "hits = sine[sine['eval'] < 1e-20].copy()\n",
    "hits['gene'] = hits.apply(hit_to_locus, axis=1, gff=gff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>target</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>ali_len</th>\n",
       "      <th>no_mism</th>\n",
       "      <th>no_go</th>\n",
       "      <th>q_start</th>\n",
       "      <th>q_end</th>\n",
       "      <th>t_start</th>\n",
       "      <th>t_end</th>\n",
       "      <th>eval</th>\n",
       "      <th>bit</th>\n",
       "      <th>qlen</th>\n",
       "      <th>species</th>\n",
       "      <th>symbol</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T_tridentatus_Six3/6</td>\n",
       "      <td>pseudochrom_3</td>\n",
       "      <td>0.983</td>\n",
       "      <td>180</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>7711226</td>\n",
       "      <td>7711047</td>\n",
       "      <td>1.348000e-35</td>\n",
       "      <td>131</td>\n",
       "      <td>60</td>\n",
       "      <td>T tridentatus</td>\n",
       "      <td>Six3/6</td>\n",
       "      <td>g5264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>P_tepidariorum_Six1/2</td>\n",
       "      <td>pseudochrom_3</td>\n",
       "      <td>0.728</td>\n",
       "      <td>177</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>7711226</td>\n",
       "      <td>7711050</td>\n",
       "      <td>2.638000e-24</td>\n",
       "      <td>98</td>\n",
       "      <td>60</td>\n",
       "      <td>P tepidariorum</td>\n",
       "      <td>Six1/2</td>\n",
       "      <td>g5264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>T_antipodiana_Six1/2</td>\n",
       "      <td>pseudochrom_3</td>\n",
       "      <td>0.694</td>\n",
       "      <td>177</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>7711226</td>\n",
       "      <td>7711050</td>\n",
       "      <td>1.772000e-23</td>\n",
       "      <td>96</td>\n",
       "      <td>60</td>\n",
       "      <td>T antipodiana</td>\n",
       "      <td>Six1/2</td>\n",
       "      <td>g5264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>A_bruennichi_Six3/6</td>\n",
       "      <td>pseudochrom_3</td>\n",
       "      <td>0.983</td>\n",
       "      <td>180</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>7711226</td>\n",
       "      <td>7711047</td>\n",
       "      <td>4.787000e-35</td>\n",
       "      <td>129</td>\n",
       "      <td>60</td>\n",
       "      <td>A bruennichi</td>\n",
       "      <td>Six3/6</td>\n",
       "      <td>g5264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>C_sculpturatus_Six3/6</td>\n",
       "      <td>pseudochrom_3</td>\n",
       "      <td>1.000</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>7711226</td>\n",
       "      <td>7711047</td>\n",
       "      <td>1.348000e-35</td>\n",
       "      <td>131</td>\n",
       "      <td>60</td>\n",
       "      <td>C sculpturatus</td>\n",
       "      <td>Six3/6</td>\n",
       "      <td>g5264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>S_maritima_Six1/2</td>\n",
       "      <td>pseudochrom_3</td>\n",
       "      <td>0.728</td>\n",
       "      <td>177</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>7711226</td>\n",
       "      <td>7711050</td>\n",
       "      <td>3.931000e-25</td>\n",
       "      <td>101</td>\n",
       "      <td>60</td>\n",
       "      <td>S maritima</td>\n",
       "      <td>Six1/2</td>\n",
       "      <td>g5264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>T_castaneum_Six1/2</td>\n",
       "      <td>pseudochrom_3</td>\n",
       "      <td>0.711</td>\n",
       "      <td>177</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>7711226</td>\n",
       "      <td>7711050</td>\n",
       "      <td>7.414000e-25</td>\n",
       "      <td>100</td>\n",
       "      <td>60</td>\n",
       "      <td>T castaneum</td>\n",
       "      <td>Six1/2</td>\n",
       "      <td>g5264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>588</th>\n",
       "      <td>A_bruennichi_Six3/6</td>\n",
       "      <td>pseudochrom_3</td>\n",
       "      <td>1.000</td>\n",
       "      <td>180</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>7711226</td>\n",
       "      <td>7711047</td>\n",
       "      <td>1.348000e-35</td>\n",
       "      <td>131</td>\n",
       "      <td>60</td>\n",
       "      <td>A bruennichi</td>\n",
       "      <td>Six3/6</td>\n",
       "      <td>g5264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>622</th>\n",
       "      <td>A_bruennichi_Six1/2</td>\n",
       "      <td>pseudochrom_3</td>\n",
       "      <td>0.694</td>\n",
       "      <td>177</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>7711226</td>\n",
       "      <td>7711050</td>\n",
       "      <td>3.624000e-24</td>\n",
       "      <td>98</td>\n",
       "      <td>60</td>\n",
       "      <td>A bruennichi</td>\n",
       "      <td>Six1/2</td>\n",
       "      <td>g5264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>A_bruennichi</td>\n",
       "      <td>pseudochrom_3</td>\n",
       "      <td>0.677</td>\n",
       "      <td>177</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>7711226</td>\n",
       "      <td>7711050</td>\n",
       "      <td>2.433000e-23</td>\n",
       "      <td>96</td>\n",
       "      <td>60</td>\n",
       "      <td>A bruennichi</td>\n",
       "      <td></td>\n",
       "      <td>g5264</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>81 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     query         target  seq_id  ali_len  no_mism  no_go  \\\n",
       "0     T_tridentatus_Six3/6  pseudochrom_3   0.983      180        1      0   \n",
       "26   P_tepidariorum_Six1/2  pseudochrom_3   0.728      177       16      0   \n",
       "34    T_antipodiana_Six1/2  pseudochrom_3   0.694      177       18      0   \n",
       "40     A_bruennichi_Six3/6  pseudochrom_3   0.983      180        1      0   \n",
       "44   C_sculpturatus_Six3/6  pseudochrom_3   1.000      180        0      0   \n",
       "..                     ...            ...     ...      ...      ...    ...   \n",
       "574      S_maritima_Six1/2  pseudochrom_3   0.728      177       16      0   \n",
       "581     T_castaneum_Six1/2  pseudochrom_3   0.711      177       17      0   \n",
       "588    A_bruennichi_Six3/6  pseudochrom_3   1.000      180        0      0   \n",
       "622    A_bruennichi_Six1/2  pseudochrom_3   0.694      177       18      0   \n",
       "631           A_bruennichi  pseudochrom_3   0.677      177       19      0   \n",
       "\n",
       "     q_start  q_end  t_start    t_end          eval  bit  qlen  \\\n",
       "0          1     60  7711226  7711047  1.348000e-35  131    60   \n",
       "26         1     59  7711226  7711050  2.638000e-24   98    60   \n",
       "34         1     59  7711226  7711050  1.772000e-23   96    60   \n",
       "40         1     60  7711226  7711047  4.787000e-35  129    60   \n",
       "44         1     60  7711226  7711047  1.348000e-35  131    60   \n",
       "..       ...    ...      ...      ...           ...  ...   ...   \n",
       "574        1     59  7711226  7711050  3.931000e-25  101    60   \n",
       "581        1     59  7711226  7711050  7.414000e-25  100    60   \n",
       "588        1     60  7711226  7711047  1.348000e-35  131    60   \n",
       "622        1     59  7711226  7711050  3.624000e-24   98    60   \n",
       "631        1     59  7711226  7711050  2.433000e-23   96    60   \n",
       "\n",
       "            species  symbol   gene  \n",
       "0     T tridentatus  Six3/6  g5264  \n",
       "26   P tepidariorum  Six1/2  g5264  \n",
       "34    T antipodiana  Six1/2  g5264  \n",
       "40     A bruennichi  Six3/6  g5264  \n",
       "44   C sculpturatus  Six3/6  g5264  \n",
       "..              ...     ...    ...  \n",
       "574      S maritima  Six1/2  g5264  \n",
       "581     T castaneum  Six1/2  g5264  \n",
       "588    A bruennichi  Six3/6  g5264  \n",
       "622    A bruennichi  Six1/2  g5264  \n",
       "631    A bruennichi          g5264  \n",
       "\n",
       "[81 rows x 16 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hits[hits['target'] == 'pseudochrom_3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "candidates = hits['gene'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "plit_peptides = {}\n",
    "with open('/Volumes/project/pycnogonum/genome/draft/transcripts.fa.transdecoder.pep', 'r') as f:\n",
    "    lines = f.readlines()\n",
    "    for i, line in enumerate(lines):\n",
    "        if line.startswith(\">\"):\n",
    "            gene = line.split(\" \")[0][1:]\n",
    "            plit_peptides[gene] = \"\"\n",
    "        else:\n",
    "            plit_peptides[gene] += line.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "candidate_sequences = {}\n",
    "for gene in plit_peptides:\n",
    "    for c in candidates:\n",
    "        if c in gene:\n",
    "            candidate_sequences[gene] = plit_peptides[gene]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Volumes/scratch/pycnogonum/genome/draft/sine/sine_candidates.fa', 'w') as f:\n",
    "    for gene, seq in candidate_sequences.items():\n",
    "        f.write(f\">{gene}\\n{seq}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can concatenate all the sequences and align them.\n",
    "\n",
    "```bash\n",
    "cat sine_chelicerates.fa > sine.fa\n",
    "cat sine_candidates.fa >> sine.fa\n",
    "```\n",
    "\n",
    "Before continuing, check whether the sequences were concatenated correctly - if there is no newline\n",
    "character at the end of `sine.fa`, the header of the first candidate sequence will be appended to\n",
    "the last sequence of the `sine_chelicerates.fa` file.\n",
    "\n",
    "First, quick-and-dirty alignment so we can just cut out the homeodomain regions from the _P.\n",
    "litorale_ sequences:\n",
    "\n",
    "```bash\n",
    "cd /lisc/scratch/zoology/pycnogonum/genome/draft/sine\n",
    "mafft --thread 4 sine.fa > sine_aligned.fasta\n",
    "```\n",
    "\n",
    "Now we can extract the homeodomain regions. This was done manually in Jalview, by deleting columns\n",
    "left and right of the homeobox region. We excluded the \"sine-like\" sequences from the alignment, as\n",
    "they visibly did not agree with the rest of the sequences.\n",
    "\n",
    "The resulting trimmed alignment was saved as `homeobox.fasta` and re-aligned with MAFFT in L-INS-i\n",
    "mode (Probably most accurate, a little bit slower):\n",
    "\n",
    "```bash\n",
    "mafft --thread 4 --maxiterate 1000 --localpair homeobox.fasta > homeobox_aligned.fasta\n",
    "```\n",
    "\n",
    "The resulting alignment was input to IQTREE2.\n",
    "\n",
    "```bash\n",
    "module load iqtree\n",
    "/usr/bin/time iqtree2 -s ./homeobox_aligned.fasta -B 1000 -T 4\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PB.8648</th>\n",
       "      <td>pseudochrom_56</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>4420024</td>\n",
       "      <td>4432656</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.8648;function=Sequence-specific DNA bind...</td>\n",
       "      <td>Six4/5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g5264</th>\n",
       "      <td>pseudochrom_3</td>\n",
       "      <td>AUGUSTUS</td>\n",
       "      <td>gene</td>\n",
       "      <td>7701237</td>\n",
       "      <td>7711655</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=g5264;function=Transcriptional regulator, S...</td>\n",
       "      <td>Six3/6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PB.2234</th>\n",
       "      <td>pseudochrom_13</td>\n",
       "      <td>PacBio</td>\n",
       "      <td>gene</td>\n",
       "      <td>4947930</td>\n",
       "      <td>4972081</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PB.2234;function=sequence-specific DNA bind...</td>\n",
       "      <td>Six1/2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  seqid    source  type    start      end score strand phase  \\\n",
       "gene                                                                           \n",
       "PB.8648  pseudochrom_56    PacBio  gene  4420024  4432656     .      +     .   \n",
       "g5264     pseudochrom_3  AUGUSTUS  gene  7701237  7711655     .      -     .   \n",
       "PB.2234  pseudochrom_13    PacBio  gene  4947930  4972081     .      +     .   \n",
       "\n",
       "                                                attributes    name  \n",
       "gene                                                                \n",
       "PB.8648  ID=PB.8648;function=Sequence-specific DNA bind...  Six4/5  \n",
       "g5264    ID=g5264;function=Transcriptional regulator, S...  Six3/6  \n",
       "PB.2234  ID=PB.2234;function=sequence-specific DNA bind...  Six1/2  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names = {\n",
    "    'PB.8648': 'Six4/5',\n",
    "    'g5264': 'Six3/6',\n",
    "    'PB.2234': 'Six1/2',\n",
    "}\n",
    "\n",
    "sine_cluster = gff.set_index('gene').loc[names.keys()]\n",
    "sine_cluster['name'] = sine_cluster.index.map(names)\n",
    "sine_cluster.sort_values(['seqid', 'start'], ascending=[False, True])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ascc24",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
